### PR TITLE
TensorIterator: Reduce serial_for_each static overhead

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58909 TensorIterator: Reduce serial_for_each static overhead**
* #58656 Replace `dim_apply` with `TensorIterator`

Differential Revision: [D28776507](https://our.internmc.facebook.com/intern/diff/D28776507)

### CODE REVIEW

**Comment by peterbell10** [comment]
I get `serial_for_each` overall instruction count dropping from 17490 to 10800, as measured with this benchmark:
```
import timeit
from torch.utils.benchmark import Timer
timer = Timer("m1.add_(b);", setup="at::Tensor m1=torch::empty({1}); at:
:Tensor b = torch::empty({1});", language="c++", timer=timeit.default_ti
mer)
stats=timer.collect_callgrind(number=30, repeats=3)
counts = stats[1].as_standardized().stats(inclusive=True)
print(counts.filter(lambda n: 'serial_for_each' in n))
```

File: aten/src/ATen/TensorIterator.cpp [hunk 1]
```diff
@@ -653,20 +670,30 @@ void TensorIteratorBase::for_each(loop2d_t loop, int64_t grain_size) {
...
+  const auto ndim = this->ndim();
+
+  c10::SmallBuffer<char*, 4> ptrs(ntensors);
+  c10::SmallBuffer<int64_t, 8> strides(ntensors * std::max(2, ndim));
+
+  at::get_base_ptrs(ptrs.data(), operands_);
+  at::get_strides(strides.data(), operands_, ndim);
+  if (ndim < 2) {
+    std::fill(strides.begin() + ntensors * ndim, strides.end(), 0);
+  }
```
Comment (peterbell10) [inline aten/src/ATen/TensorIterator.cpp:693]: Zero-padding here costs around 800 instructions, so still not free but still cheaper than the way it was done previously.

**Comment by peterbell10** [comment]
@ngimel on master, after your PR, the same benchmark gives 14310 so this is a further improvement.

**Comment by ngimel** [comment]
fyi @bhosmer, @ezyang

File: aten/src/ATen/TensorIterator.cpp [hunk 2]
```diff
@@ -653,20 +670,30 @@ void TensorIteratorBase::for_each(loop2d_t loop, int64_t grain_size) {
...
+  const auto ntensors = this->ntensors();
+  const auto ndim = this->ndim();
+
+  c10::SmallBuffer<char*, 4> ptrs(ntensors);
+  c10::SmallBuffer<int64_t, 8> strides(ntensors * std::max(2, ndim));
+
+  at::get_base_ptrs(ptrs.data(), operands_);
+  at::get_strides(strides.data(), operands_, ndim);
+  if (ndim < 2) {
+    std::fill(strides.begin() + ntensors * ndim, strides.end(), 0);
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/TensorIterator.cpp:692]: Currently I'm preserving the `TensorIterator` public methods with the same return types and values as before. So, `TensorIterator.get_strides()` doesn't pad with zeros. If it's okay to removed the public method then I could move the zero-padding into `get_strides` though.

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/TensorIterator.cpp:683]: It's ok to change it (we are changing add_input and add_output for example)

**Comment by ezyang** [comment]
All for performance optimizations; changes to utility code LGTM, I didn't carefully review the substantive logic, holler if you need more eyes.

**Comment by ngimel** [comment]
This looks good, can you please post final benchmarks?

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28776507).

**Comment by peterbell10** [comment]
The final callgrind result was 10650 instructions in `serial_for_each`.

File: aten/src/ATen/TensorIterator.cpp [hunk 3]
```diff
@@ -18,6 +21,29 @@ using PtrVector = TensorIteratorBase::PtrVector;
...
+inline void get_base_ptrs(char** ptrs, ArrayRef<OperandInfo> operands) {
+  std::transform(operands.begin(), operands.end(), ptrs, [](const OperandInfo& op) {
+    return static_cast<char*>(op.data);
+  });
+}
+
+inline void get_strides(int64_t* strides, ArrayRef<OperandInfo> operands, int64_t ndim) {
+  for (int64_t dim = 0; dim < ndim; ++dim) {
+    for (size_t arg = 0; arg < operands.size(); ++arg) {
+      *strides++ = operands[arg].stride_bytes[dim];
```
Comment (ngimel) [inline aten/src/ATen/TensorIterator.cpp:35]:
```suggestion
    for (const auto & operand : operands) {
      *strides++ = operand.stride_bytes[dim];
```

Comment (ngimel) [reply aten/src/ATen/TensorIterator.cpp:35]: I've left a suggestion from internal linter here, commit it directly if you want

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28776507).
